pkg/authz/oauth/authclient.go

```diff
@@ -0,0 +1,177 @@
+func NewAuthClient(tokenURL string, apiClient api.Client, opts ...AuthClientOption) (AuthClient, error) {
+	client := &authClient{
+		tokenURL:      tokenURL,
+		apiClient:     apiClient,
+		getTokenMutex: &sync.Mutex{},
+		options:       &authClientOptions{},
+	}
+	for _, o := range opts {
+		o(client.options)
+	}
+
+	if client.options.serverName == "" {
+		client.options.serverName = defaultServerName
+	}
+	if client.options.authenticator == nil {
+		return nil, errors.New("unable to create client, no authenticator configured")
+	}
+	return client, nil
+}
+
+// WithServerName - sets up the server name in auth client
+func WithServerName(serverName string) AuthClientOption {
+	return func(opt *authClientOptions) {
+		opt.serverName = serverName
+	}
+}
+
+// WithClientSecretAuth - sets up to use client secret authenticator
+func WithClientSecretAuth(clientID, clientSecret string) AuthClientOption {
+	return func(opt *authClientOptions) {
+		opt.authenticator = &clientSecretAuthenticator{
+			clientID,
+			clientSecret,
+		}
+	}
+}
+
+// WithKeyPairAuth - sets up to use public/private key pair authenticator
+func WithKeyPairAuth(clientID, audience string, privKey *rsa.PrivateKey, publicKey []byte) AuthClientOption {
+	return func(opt *authClientOptions) {
+		opt.authenticator = &keyPairAuthenticator{
+			clientID,
+			audience,
+			privKey,
+			publicKey,
+		}
+	}
+}
+
+func (c *authClient) getCachedToken() string {
+	if c.cachedToken != nil {
+		select {
+		case <-c.cachedTokenExpiry.C:
+			// cleanup the token on expiry
+			c.cachedToken = nil
+			return ""
+		default:
+			return c.cachedToken.AccessToken
+		}
+	}
+	return ""
+}
+
+// GetToken returns a token from cache if not expired or fetches a new token
+func (c *authClient) GetToken() (string, error) {
+	// only one GetToken should execute at a time
+	c.getTokenMutex.Lock()
+	defer c.getTokenMutex.Unlock()
+
+	if token := c.getCachedToken(); token != "" {
+		return token, nil
+	}
+
+	// try fetching a new token
+	return c.fetchNewToken()
+}
+
+// fetchNewToken fetches a new token from the platform and updates the token cache.
+func (c *authClient) fetchNewToken() (string, error) {
+	tokenResponse, err := c.getOAuthTokens()
+	if err != nil {
+		return "", err
+	}
+
+	almostExpires := (tokenResponse.ExpiresIn * 4) / 5
+
+	c.cachedToken = tokenResponse
+	c.cachedTokenExpiry = time.NewTimer(time.Duration(almostExpires) * time.Second)
+	return c.cachedToken.AccessToken, nil
+}
+
+func (c *authClient) getOAuthTokens() (*tokenResponse, error) {
+	startTime := time.Now()
+	req, err := c.options.authenticator.prepareRequest()
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.postAuthForm(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.Code != 200 {
+		body := resp.Body
+		log.Debugf("bad response from %s: %d %s: %s, request time : %s", c.options.serverName, resp.Code, http.StatusText(resp.Code), body, startTime.String())
```

Comment:
Can you update to use the FieldLogger please

Reply:
Sure